File: C++/255.verify-preorder-sequence-in-binary-search-tree.cpp
/*
 * @lc app=leetcode id=255 lang=cpp
 *
 * [255] Verify Preorder Sequence in Binary Search Tree 
 * 
 * Given an array of numbers, verify whether it is the correct preorder traversal sequence of a binary search tree.
 * 
 * You may assume each number in the sequence is unique.
 * 
 * Consider the following binary search tree: 
 *      5
 *     / \
 *    2   6
 *   / \
 *  1   3
 * Example 1:
 * 
 * Input: [5,2,6,1,3]
 * Output: false
 * 
 * Example 2: 
 * Input: [5,2,1,3,6]
 * Output: true
 * Follow up:
 * Could you do it using only constant space complexity?
 * 
 * 
 *  
 */

#include<vector>
#include<stack>
#include<queue>
#include<stack>
#include<iostream>

using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// @lc code=start
class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) {
        int root = INT_MIN;
        vector<int> stack;
        for (int num : preorder) {
            if (num < root) return false;

            while (!stack.empty() && num > stack.back()) {
                root = stack.back(); 
                stack.pop_back();
            }
            stack.push_back(num);
        }
        return true;
    }
    
    bool verifyPostorder(vector<int>& postorder) {
        vector<int> stack;

        int pre_elem = INT_MAX;
        for(int i = postorder.size() - 1; i >= 0; --i) {
            if(postorder[i] > pre_elem) return false;

            while(!stack.empty() && postorder[i] < stack.back()) {
                pre_elem = stack.back();
                stack.pop_back();
            }
            stack.push_back(postorder[i]);
        }
        return true;
    }


    bool verifyPreorder(vector<int>& preorder) {
        return DFS(preorder, 0, preorder.size()-1);
    }
    bool DFS(vector<int>& preorder, int left, int right) {
        if (left >= right) return true;

        int root = preorder[left];
       
        int idx = left+1;
        while (preorder[idx] < root) idx++;

        int tmp = idx;
        while (tmp < right) {
            if (preorder[tmp++] < root) 
                return false;
        }

        return DFS(preorder, left+1, idx-1) && 
               DFS(preorder, idx, right);
    }

};

// @lc code=end

File: Python/255.verify-preorder-sequence-in-binary-search-tree.py
#
# @lc app=leetcode id=255 lang=python3
#
# [255] Verify Preorder Sequence in Binary Search Tree
#
# https://leetcode.com/problems/verify-preorder-sequence-in-binary-search-tree/description/
#
# algorithms
# Medium (49.95%)
# Likes:    1201
# Dislikes: 89
# Total Accepted:    81.7K
# Total Submissions: 163.2K
# Testcase Example:  '[5,2,1,3,6]'
#
# Given an array of unique integers preorder, return true if it is the correct
# preorder traversal sequence of a binary search tree.
# 
# 
# Example 1:
# 
# 
# Input: preorder = [5,2,1,3,6]
# Output: true
# 
# 
# Example 2:
# 
# 
# Input: preorder = [5,2,6,1,3]
# Output: false
# 
# 
# 
# Constraints:
# 
# 
# 1 <= preorder.length <= 10^4
# 1 <= preorder[i] <= 10^4
# All the elements of preorder are unique.
# 
# 
# 
# Follow up: Could you do it using only constant space complexity?
# 
#

# @lc code=start
class Solution:
    def verifyPreorder(self, preorder: List[int]) -> bool:
        min_val = float('-inf')
        stack = []
        for num in preorder:
            while stack and num > stack[-1]:
                min_val = stack.pop()
            if num < min_val:
                return False
            stack.append(num)
        return True
    
        inorder = []
        stack = []
        for num in preorder:
            while stack and num > stack[-1]:
                inorder.append(stack.pop())
            stack.append(num)
        inorder.extend(reversed(stack))
        return inorder == sorted(inorder)
        
# @lc code=end

